### DISCUSSION TITLE
`.clang-tidy` config file doesn't work

### DISCUSSION THREAD

**Discussion by demberto** [microsoft/vscode-cpptools#9759] [Q&A]
Before going ahead and opening an issue, I want to confirm whether `.clang-tidy` doesn't work with a hierarchy like this:

```
├─── .vscode
├─── build
├─── .clang-tidy
├─── src.c
└─── src.h
```

This is what `.clang-tidy` has
```yaml
---
Checks: '-clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling'
---
```
This doesn't work, but adding it to project wide `settings.json` like this works:

```json
{
    "C_Cpp.codeAnalysis.clangTidy.checks.disabled": ["clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling"]
}
```

**Comment by sean-mcmanus** [comment]
The .clang-tidy (or .clang-format) file is only picked up at the sibling or parent level in relation to a source file, so if you have a .clang-tidy file in some subfolder, it won't get picked up. The detection of the .clang-tidy is done by the clang-tidy executable itself and not our extension (the only spot where our extension check for .clang-tidy when determining when the `"C_Cpp.codeAnalysis.clangTidy.fallbackConfig"` setting should be used).

**Reply by sean-mcmanus** [reply]
Oh, if .clang-tidy and src.c are in the same folder, it should usually be used -- can you provide more bug repro info?

**Reply by sean-mcmanus** [reply]
I might be reproing the issue you're seeing, i.e. I'm seeing check disabling via the .clang-tidy file not working. I'm investigating...

**Reply by sean-mcmanus** [reply]
There are multiple issues:
1. You need to make sure you don't have C_Cpp.codeAnalysis.clangTidy.checks.enabled set to `"*"` or it will add that after the .clang-tidy checks, e.g. the same as if you were using `Checks: '-readability-redundant-declaration,*` (the '*' re-adds the excluded check). The description for that setting mentions that the setting values are appended.
2. You need to remove the ending --- from you .clang-tidy file.
e.g. using
```
---
Checks: '*,-readability-redundant-declaration'
```
works.

**Reply by sean-mcmanus** [reply]
Actually, the end --- is fine -- it just requires a newline after it (or it'll output an error message).

**Reply by demberto** [reply]
![image](https://user-images.githubusercontent.com/65948403/185708310-f08ab979-46a7-4c04-905b-5cc3772b6239.png)
The setting you mentioned is empty. I can reproduce the same issue if I make subdirectories for the source code, and keep `.clang-tidy`. What really makes me think its an extension only issue is that configuring via `C_Cpp.codeAnalysis.clangTidy.checks.disabled` works perfectly.

**Reply by demberto** [reply]
Something is really messed up, even the `codeAnalysis` exclusion doesn't work either
![image](https://user-images.githubusercontent.com/65948403/185709772-2868f504-df91-4944-89c0-177ea0713a39.png)

**Reply by sean-mcmanus** [reply]
I see `clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling` not working at all in the .clang-tidy file or the disabled setting and I need to use `clang-diagnostic-deprecated-declarations`.

I'm seeing the C_Cpp.codeAnalysis.exclude setting working. I'm not sure what could be causing it to not work -- what OS are you on? There might be some issue with the exclude regex not matching a path for some reason, such as having a symlink on the path.

**Reply by demberto** [reply]
I am on Windows and external contains submodules, I tried quite a few combinations, none of them worked. Should I open an issue for the `.clang-tidy` file (in clang-tidy repo)?

**Reply by sean-mcmanus** [reply]
If you think there is a bug with our extension's handling of the .clang-tidy file you should file a bug on our extension...otherwise if you can get a repro via running clang-tidy from the command line then you should report the bug on the LLVM GitHub page.
